agent/src/apiclient.rs

```diff
@@ -0,0 +1,296 @@
+    let mut raw_args = Vec::new();
+    raw_args.push("/updates/status".to_string());
+
+    let update_status_output = invoke_apiclient(get_raw_args(raw_args))?;
+
+    let update_status_string = String::from_utf8_lossy(&update_status_output.stdout).to_string();
+    let update_status: UpdateStatus =
+        serde_json::from_str(&update_status_string).context(UpdateStatusContent)?;
+
+    Ok(update_status)
+}
+
+/// extract OS info from the local system. For now this is just the version id.
+pub fn get_os_info() -> Result<OsInfo> {
+    let mut raw_args = Vec::new();
+    raw_args.push("/os".to_string());
+
+    let os_info_output = invoke_apiclient(get_raw_args(raw_args))?;
+
+    let os_info_content_string = String::from_utf8_lossy(&os_info_output.stdout).to_string();
+    let os_info_content_json: serde_json::Value =
+        serde_json::from_str(&os_info_content_string).context(OsContent)?;
+
+    Ok(OsInfo {
+        version_id: os_info_content_json["version_id"].to_string(),
+    })
+}
+
+pub fn get_chosen_update() -> Result<Option<UpdateImage>> {
+    let update_status = get_update_status()?;
+
+    Ok(update_status.chosen_update)
+}
+
+pub fn get_most_recent_command() -> Result<CommandResult> {
+    let update_status = get_update_status()?;
+
+    Ok(update_status.most_recent_command)
+}
+
+pub fn refresh_updates() -> Result<()> {
+    let mut raw_args = Vec::new();
+    raw_args.push("/actions/refresh-updates".to_string());
+    raw_args.push("-m".to_string());
+    raw_args.push("POST".to_string());
+
+    invoke_apiclient(raw_args)?;
+
+    Ok(())
+}
+
+pub fn prepare_update() -> Result<()> {
+    let mut raw_args = Vec::new();
+    raw_args.push("/actions/prepare-update".to_string());
+    raw_args.push("-m".to_string());
+    raw_args.push("POST".to_string());
+
+    invoke_apiclient(raw_args)?;
+
+    Ok(())
+}
+
+pub fn activate_update() -> Result<()> {
+    let mut raw_args = Vec::new();
+    raw_args.push("/actions/activate-update".to_string());
+    raw_args.push("-m".to_string());
+    raw_args.push("POST".to_string());
+
+    invoke_apiclient(raw_args)?;
+
+    Ok(())
+}
+
+pub fn reboot() -> Result<Output> {
+    let mut raw_args = Vec::new();
+    raw_args.push("/actions/reboot".to_string());
+    raw_args.push("-m".to_string());
+    raw_args.push("POST".to_string());
+
+    Ok(invoke_apiclient(raw_args)?)
+}
+
+// List all available versions which current Bottlerocket OS can update to.
+pub fn list_available() -> Result<Vec<String>> {
+    // Refresh list of updates and check if there are any available
+    refresh_updates()?;
+
+    let update_status = get_update_status()?;
+
+    // Raise error if failed to refresh update or update acton performed out of band
+    if update_status.most_recent_command.cmd_type != UpdateCommand::Refresh
+        && update_status.most_recent_command.cmd_status != CommandStatus::Success
+    {
+        return RefreshUpdate {}.fail();
+    }
```

Comment:
You can include `snafu::ensure` and do

```suggestion
   ensure!(
      update_status.most_recent_command.cmd_type != UpdateCommand::Refresh 
      && update_status.most_recent_command.cmd_status != CommandStatus::Success, 
      RefreshUpdate
   );
  
```